pkg/iacinventory/github_repo.go

```diff
@@ -0,0 +1,184 @@
+	if !info.Mode().IsRegular() {
+		return nil
+	}
+	// If the file ends with TF, the parent directory contains terraform files
+	if strings.HasSuffix(info.Name(), ".tf") {
+		if g.TerraformDirs == nil {
+			g.TerraformDirs = make(map[string]bool)
+		}
+		g.TerraformDirs[strings.TrimPrefix(filepath.Dir(filepath.Clean(path)), g.dir+"/")] = true
+	}
+	return nil
+}
+
+// GetCloudFormationDirs implements WalkFunc to search for CloudFormation files in a repository.
+func (g *GithubRepo) GetCloudFormationDirs(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	// if it is anything other than a regular file, return
+	if !info.Mode().IsRegular() {
+		return nil
+	}
+	// Cloudformation files do not have a unique extension, and are *typically*
+	// ".yaml" or ".json" by convention. However, sometimes organizations use
+	// Jinja, Go, or some other utility to template their Cloudformation.
+	//
+	// If the file has a possible extension and contains the string 'AWSTemplateFormatVersion',
+	// then it is *most likely* a CF file.
+
+	const maxSizeForCloudFormationConsideration int64 = 5 << (10 * 2) // 5MB, which is VERY large for json/yaml data
+	if info.Size() > maxSizeForCloudFormationConsideration {
+		// Exit early - file disqualified due to size.
+		return nil
+	}
+
+	if strings.HasSuffix(info.Name(), ".json") ||
+		strings.HasSuffix(info.Name(), ".yaml") ||
+		strings.HasSuffix(info.Name(), ".yml") ||
+		strings.HasSuffix(info.Name(), ".template") {
+		f, err := os.Open(filepath.Clean(path))
+		if err != nil {
+			return fmt.Errorf("error opening file during CloudFormation analysis: %w", err)
+		}
+		scanner := bufio.NewScanner(f)
+		for scanner.Scan() {
+			if bytes.Contains(scanner.Bytes(), []byte("AWSTemplateFormatVersion")) {
+				if g.CloudformationDirs == nil {
+					g.CloudformationDirs = make(map[string]bool)
+				}
+				g.CloudformationDirs[strings.TrimPrefix(filepath.Dir(filepath.Clean(path)), g.dir+"/")] = true
+			}
+		}
+		f.Close()
+	}
+	return nil
+}
+
+// GetMaster fetches and extracts the tarball of the master branch.
+func (g *GithubRepo) GetMaster(ctx context.Context, c *http.Client, username string, oauthToken string) error {
+	if username == "" || oauthToken == "" {
+		return fmt.Errorf("error fetching repository: credentials are not set")
+	}
+	req, _ := http.NewRequestWithContext(ctx, "GET", "https://api.github.com/repos/"+g.FullName+"/tarball/master", nil)
+	req.Header.Set("Authorization", "token "+oauthToken)
+	req.Header.Set("Accept", "application/vnd.github.v3.raw")
+	resp, err := c.Do(req)
+	if err != nil {
+		return fmt.Errorf("error sending HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// we stuff the extracted tarball into a temporary directory
+	dir, err := ioutil.TempDir("", "soluble-iac-scan")
+	if err != nil {
+		return fmt.Errorf("error creating temporary directory: %w", err)
+	}
+	g.dir = dir
+	gunzip, err := gzip.NewReader(resp.Body)
```

Comment:
Let's move this gunzip/untar code to a new package and include a test case.

We would add unzip code there as well.